File: src/features/verify/verifyRepository.js
const db = require('../../database/db');

// --- Guild config: which role to give / (optionally) remove for each verification
// type, plus the channel where verification reports get posted ---

async function getGuildConfig(guildId) {
  await db.ready;
  const result = await db.client.execute({
    sql: 'SELECT * FROM verify_role_config WHERE guild_id = ?',
    args: [guildId],
  });

  const row = result.rows[0];
  return row
    ? {
        guild_id: row.guild_id,
        sub_give_role_id: row.sub_give_role_id,
        sub_remove_role_id: row.sub_remove_role_id,
        domme_give_role_id: row.domme_give_role_id,
        domme_remove_role_id: row.domme_remove_role_id,
        maledom_give_role_id: row.maledom_give_role_id,
        maledom_remove_role_id: row.maledom_remove_role_id,
        report_channel_id: row.report_channel_id,
      }
    : {
        guild_id: guildId,
        sub_give_role_id: null,
        sub_remove_role_id: null,
        domme_give_role_id: null,
        domme_remove_role_id: null,
        maledom_give_role_id: null,
        maledom_remove_role_id: null,
        report_channel_id: null,
      };
}

// Always writes all 7 columns (the manager merges with the existing row first, so
// callers never need to worry about accidentally clearing a value that wasn't touched).
async function setGuildConfig(guildId, fields) {
  await db.ready;
  await db.client.execute({
    sql: `INSERT INTO verify_role_config
            (guild_id, sub_give_role_id, sub_remove_role_id, domme_give_role_id, domme_remove_role_id, maledom_give_role_id, maledom_remove_role_id, report_channel_id)
          VALUES (?, ?, ?, ?, ?, ?, ?, ?)
          ON CONFLICT(guild_id) DO UPDATE SET
            sub_give_role_id = excluded.sub_give_role_id,
            sub_remove_role_id = excluded.sub_remove_role_id,
            domme_give_role_id = excluded.domme_give_role_id,
            domme_remove_role_id = excluded.domme_remove_role_id,
            maledom_give_role_id = excluded.maledom_give_role_id,
            maledom_remove_role_id = excluded.maledom_remove_role_id,
            report_channel_id = excluded.report_channel_id`,
    args: [
      guildId,
      fields.sub_give_role_id,
      fields.sub_remove_role_id,
      fields.domme_give_role_id,
      fields.domme_remove_role_id,
      fields.maledom_give_role_id,
      fields.maledom_remove_role_id,
      fields.report_channel_id,
    ],
  });
}

module.exports = {
  getGuildConfig,
  setGuildConfig,
};

File: discord-bot-python/src/features/verify/verify_repository.py
from src.database import db

# --- Guild config (which roles to assign, where to post reports) ---


async def get_guild_config(guild_id: str) -> dict:
    await db.ready()
    result = await db.client.execute("SELECT * FROM verify_guild_config WHERE guild_id = ?", [guild_id])
    rows = db.rows_as_dicts(result)

    if rows:
        row = rows[0]
        return {
            "guild_id": row["guild_id"],
            "findom_role_id": row["findom_role_id"],
            "sub_role_id": row["sub_role_id"],
            "verified_channel_id": row["verified_channel_id"],
        }
    return {"guild_id": guild_id, "findom_role_id": None, "sub_role_id": None, "verified_channel_id": None}


async def set_findom_role(guild_id: str, role_id: str) -> None:
    await db.ready()
    await db.client.execute(
        """INSERT INTO verify_guild_config (guild_id, findom_role_id, sub_role_id, verified_channel_id)
           VALUES (?, ?, NULL, NULL)
           ON CONFLICT(guild_id) DO UPDATE SET findom_role_id = excluded.findom_role_id""",
        [guild_id, role_id],
    )


async def set_sub_role(guild_id: str, role_id: str) -> None:
    await db.ready()
    await db.client.execute(
        """INSERT INTO verify_guild_config (guild_id, findom_role_id, sub_role_id, verified_channel_id)
           VALUES (?, NULL, ?, NULL)
           ON CONFLICT(guild_id) DO UPDATE SET sub_role_id = excluded.sub_role_id""",
        [guild_id, role_id],
    )


async def set_verified_channel(guild_id: str, channel_id: str) -> None:
    await db.ready()
    await db.client.execute(
        """INSERT INTO verify_guild_config (guild_id, findom_role_id, sub_role_id, verified_channel_id)
           VALUES (?, NULL, NULL, ?)
           ON CONFLICT(guild_id) DO UPDATE SET verified_channel_id = excluded.verified_channel_id""",
        [guild_id, channel_id],
    )


# --- Verification records (one per guild + user + type) ---


async def upsert_verification(
    guild_id: str,
    user_id: str,
    type_: str,
    social: str | None,
    method: str,
    verified_at: int,
    verified_by: str,
    channel_id: str | None,
    message_id: str | None,
) -> None:
    await db.ready()
    await db.client.execute(
        """INSERT INTO verify_entries
             (guild_id, user_id, type, social, method, verified_at, verified_by, channel_id, message_id)
           VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
           ON CONFLICT(guild_id, user_id, type) DO UPDATE SET
             social = excluded.social,
             method = excluded.method,
             verified_at = excluded.verified_at,
             verified_by = excluded.verified_by,
             channel_id = excluded.channel_id,
             message_id = excluded.message_id""",
        [guild_id, user_id, type_, social, method, verified_at, verified_by, channel_id, message_id],
    )


async def get_verification(guild_id: str, user_id: str, type_: str) -> dict | None:
    await db.ready()
    result = await db.client.execute(
        "SELECT * FROM verify_entries WHERE guild_id = ? AND user_id = ? AND type = ?",
        [guild_id, user_id, type_],
    )
    rows = db.rows_as_dicts(result)
    return rows[0] if rows else None


async def update_verification_fields(guild_id: str, user_id: str, type_: str, social: str | None, method: str) -> None:
    await db.ready()
    await db.client.execute(
        "UPDATE verify_entries SET social = ?, method = ? WHERE guild_id = ? AND user_id = ? AND type = ?",
        [social, method, guild_id, user_id, type_],
    )
